providers/base/bin/pipewire_test.py

```diff
@@ -0,0 +1,608 @@
+        # contain only 2205 Hz, not 4410 Hz.
+        max_frequency = self.sampling_frequency / 2
+        if frequency > max_frequency or frequency < 0:
+            return None
+        band = float(frequency) / (max_frequency / len(self.spectrum))
+        return int(math.ceil(band)) - 1
+
+    def frequencies_for_band(self, band):
+        """Convenience function to tell me the delimiting frequencies
+           for a band
+        """
+        if band >= len(self.spectrum) or band < 0:
+            return None
+        lower = self.frequencies[band]
+        upper = lower + ((self.sampling_frequency / 2.0) / len(self.spectrum))
+        return (lower, upper)
+
+    def sampling_complete(self):
+        return self.number_of_samples >= self.wanted_samples
+
+
+class GStreamerMessageHandler(object):
+    def __init__(self, rec_level_range, logger, volumecontroller,
+                 pidcontroller, spectrum_analyzer):
+        """Initializes the message handler. It knows how to handle
+           spectrum and level gstreamer messages.
+
+           Arguments:
+           rec_level_range: tuple with acceptable recording level
+                            ranges
+           logger: logging object with debug, info, error methods.
+           volumecontroller: an instance of VolumeController to use
+                             to adjust RECORDING level
+           pidcontroller: a PID controller instance which helps control
+                          volume
+           spectrum_analyzer: instance of SpectrumAnalyzer to collect
+                              data from spectrum messages
+
+        """
+        self.current_level = sys.maxsize
+        self.logger = logger
+        self.pid_controller = pidcontroller
+        self.rec_level_range = rec_level_range
+        self.spectrum_analyzer = spectrum_analyzer
+        self.volume_controller = volumecontroller
+
+    def set_quit_method(self, method):
+        """ Method that will be called when sampling is complete."""
+        self._quit_method = method
+
+    def bus_message_handler(self, bus, message):
+        if message.type == Gst.MessageType.ELEMENT:
+            message_name = message.get_structure().get_name()
+            if message_name == 'spectrum':
+                # TODO: Due to an upstream bug, a structure's get_value method
+                # doesn't work if the value in question is an array (as is the
+                # case with the magnitudes).
+                # https://bugzilla.gnome.org/show_bug.cgi?id=693168
+                # We have to resort to parsing the string representation of the
+                # structure. It's an ugly hack but it works.
+                # Ideally we'd be able to say this to get fft_magnitudes:
+                # message.get_structure.get_value('magnitude').
+                # If an upstream fix ever makes it into gstreamer,
+                # remember to remove this hack and the parse_spectrum method
```

Comment:
Could you please also note this in PR description?

Reply:
I could add these, but this is from the old `audio_test.py`.

Reply:
This bug is[ fixed](https://bugzilla.gnome.org/show_bug.cgi?id=753754) in 2017, then checkbox has to include `python3-gst-1.0 (gst-python 1.12)` as dependency to solve this. Is it acceptable to add it in to checkbox ?

Reply:
`gst-python 1.12` only support ubuntu 18+, if we fix this issue properly, need to consider the ubuntu 16 case.